compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -446,12 +447,195 @@ impl<'a> Parser<'a> {
             || self.token.kind == token::CloseDelim(Delimiter::Parenthesis)
                 && self.look_ahead(1, Token::is_range_separator);
 
+        let span = expr.span;
+
         Some((
-            self.dcx().emit_err(UnexpectedExpressionInPattern { span: expr.span, is_bound }),
-            expr.span,
+            self.dcx().stash_err(
+                span,
+                StashKey::ExprInPat,
+                UnexpectedExpressionInPattern { span, is_bound },
+            ),
+            span,
         ))
     }
 
+    /// Called by [`Parser::parse_stmt_without_recovery`], used to add statement-aware subdiagnostics to the errors stashed
+    /// by [`Parser::maybe_recover_trailing_expr`].
+    pub(super) fn maybe_emit_stashed_expr_in_pats(&mut self, stmt: &Stmt) {
```

Comment:
Since stashed diagnostics do get emitted eventually — even if not stolen (to be modified or replaced) — I think this method should be renamed not to contain `emit` but `maybe_modify` or sth. like that as well as the word `suggestion` maybe. This function essentially 'goes over' all expr-in-pat diagnostics and *maybe* enhances them by potentially adding statement/context-specific suggestions.

Reply:
I opted for `maybe_augment_stashed_expr_in_pats_with_suggestions`